feat: check ARG_MAX on Unix platforms

weihanglo opened the pull request:
On Unix the limits can be gargantuan anyway so we're pretty unlikely to hit them, but might still exceed it.
We consult ARG_MAX here to get an estimate.

Fixes #138421

r? @jieyouxu

weihanglo committed:
fix(linker): prevent overflow when estimating CLI arg list length

This also updates the estimate on Windows of the length argument
list to `saturating_add` to avoid overflow.

weihanglo commented on compiler/rustc_codegen_ssa/src/back/command.rs:
```diff
@@ -137,44 +137,60 @@ impl Command {
+                .iter()
+                .map(|(k, v)| k.as_encoded_bytes().len() + 1 + v.as_encoded_bytes().len() + 1 + ptr_size)
+                .sum::<usize>();
+            let arg_max = match unsafe { libc::sysconf(libc::_SC_ARG_MAX) } {
```
This doesn't check overflow. Should we?

jieyouxu replied:
Probably use `saturing_add`. Alternatively, checked addition that just falls back to go to the OS anyway if it would overflow.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13825246795/job/38678903000) [(plain)](https://github.com/rust-lang/rust/commit/954fd05cc3e69e83bc83d2385928da6fb88fe7e3/checks/38678903000/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.941 Building wheels for collected packages: reuse
#12 2.943   Building wheel for reuse (pyproject.toml): started
#12 3.157   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.158   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=0c2fd2aaf7b0bf8d6e131220aff14712a774c2ca462f3204d25460cbcf610b63
#12 3.158   Stored in directory: /tmp/pip-ephem-wheel-cache-33t8llk2/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.161 Successfully built reuse
#12 3.161 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.578 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.579 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.148 Collecting virtualenv
#12 4.187   Downloading virtualenv-20.29.3-py3-none-any.whl (4.3 MB)
#12 4.254      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 66.8 MB/s eta 0:00:00
#12 4.327 Collecting filelock<4,>=3.12.2
#12 4.331   Downloading filelock-3.17.0-py3-none-any.whl (16 kB)
#12 4.350 Collecting distlib<1,>=0.3.7
#12 4.356   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.363      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 81.1 MB/s eta 0:00:00
#12 4.405 Collecting platformdirs<5,>=3.9.1
#12 4.410   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#12 4.495 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.696 Successfully installed distlib-0.3.9 filelock-3.17.0 platformdirs-4.3.6 virtualenv-20.29.3
#12 4.696 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.8s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      126912 kB
DirectMap2M:     7213056 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-02-18/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
Diff in /checkout/compiler/rustc_codegen_ssa/src/back/command.rs:181:
             let envs_size = self
                 .env
                 .iter()
-                .map(|(k, v)| k.as_encoded_bytes().len() + 1 + v.as_encoded_bytes().len() + 1 + ptr_size)
+                .map(|(k, v)| {
+                    k.as_encoded_bytes().len() + 1 + v.as_encoded_bytes().len() + 1 + ptr_size
+                })
                 .sum::<usize>();
             let arg_max = match unsafe { libc::sysconf(libc::_SC_ARG_MAX) } {
                 -1 => return false, // Go to OS anyway.
fmt: checked 5897 files
Build completed unsuccessfully in 0:01:19
  local time: Thu Mar 13 01:56:52 UTC 2025
  network time: Thu, 13 Mar 2025 01:56:52 GMT
##[error]Process completed with exit code 1.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13826432638/job/38682198584) [(plain)](https://github.com/rust-lang/rust/commit/b4e66b7a78fe5eaef536994dbd128f8e9951ac4b/checks/38682198584/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 19.7s done
#19 DONE 25.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
   Compiling rustc_privacy v0.0.0 (/checkout/compiler/rustc_privacy)
error[E0689]: can't call method `saturating_add` on ambiguous numeric type `{integer}`
   --> compiler/rustc_codegen_ssa/src/back/command.rs:171:55
    |
171 |                 self.args.iter().fold(0, |acc, a| acc.saturating_add(a.as_encoded_bytes().len()));
    |                                           ---         ^^^^^^^^^^^^^^
    |                                           |
    |                                           you must specify a type for this binding, like `i32`

error[E0689]: can't call method `saturating_add` on ambiguous numeric type `{integer}`
   --> compiler/rustc_codegen_ssa/src/back/command.rs:182:21
    |
179 |             let args_size = self.args.iter().fold(0, |acc, a| {
    |                                                       --- you must specify a type for this binding, like `i32`
...
182 |                 acc.saturating_add(arg).saturating_add(nul).saturating_add(ptr_size)
    |                     ^^^^^^^^^^^^^^

error[E0689]: can't call method `saturating_add` on ambiguous numeric type `{integer}`
   --> compiler/rustc_codegen_ssa/src/back/command.rs:190:21
    |
185 |             let envs_size = self.env.iter().fold(0, |acc, (k, v)| {
    |                                                      --- you must specify a type for this binding, like `i32`
...
190 |                 acc.saturating_add(k)
    |                     ^^^^^^^^^^^^^^

   Compiling rustc_passes v0.0.0 (/checkout/compiler/rustc_passes)
```

</details>

jieyouxu commented on compiler/rustc_codegen_ssa/src/back/command.rs:
```diff
@@ -137,44 +137,72 @@ impl Command {
+                -1 => return false, // Go to OS anyway.
+                max => max as usize,
+            };
+            args_size + envs_size > arg_max
```
Nit: this can also overflow, maybe also `.saturing_add` this? This entire helper is heuristics anyway and we shouldn't ICE on it.

weihanglo replied:
God, I wish we have a `/s+` operator for saturating_add
/s

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13826760501/job/38683069162) [(plain)](https://github.com/rust-lang/rust/commit/9f77ec19b5c7a0c96e920b63e06f72034265c813/checks/38683069162/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_hir_analysis v0.0.0 (/checkout/compiler/rustc_hir_analysis)
error[E0433]: failed to resolve: use of unresolved module or unlinked crate `libc`
   --> compiler/rustc_codegen_ssa/src/back/command.rs:198:42
    |
198 |             let arg_max = match unsafe { libc::sysconf(libc::_SC_ARG_MAX) } {
    |                                          ^^^^ use of unresolved module or unlinked crate `libc`
    |
    = help: if you wanted to use a crate named `libc`, use `cargo add libc` to add it to your `Cargo.toml`

error[E0433]: failed to resolve: use of unresolved module or unlinked crate `libc`
   --> compiler/rustc_codegen_ssa/src/back/command.rs:198:56
    |
198 |             let arg_max = match unsafe { libc::sysconf(libc::_SC_ARG_MAX) } {
    |                                                        ^^^^ use of unresolved module or unlinked crate `libc`
    |
    = help: if you wanted to use a crate named `libc`, use `cargo add libc` to add it to your `Cargo.toml`

    Checking rustc_mir_build v0.0.0 (/checkout/compiler/rustc_mir_build)
For more information about this error, try `rustc --explain E0433`.
error: could not compile `rustc_codegen_ssa` (lib) due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
```

</details>

jieyouxu commented:
You probably need to declare a `libc = "0.2"` in cg_ssa.

jieyouxu commented on compiler/rustc_codegen_ssa/src/back/command.rs:
```diff
@@ -137,44 +137,75 @@ impl Command {
+                -1 => return false, // Go to OS anyway.
+                max => max as usize,
+            };
+            #[cfg(not(unix))]
+            let arg_max = 0;
```
Note that this `cfg(not(unix))` is already guarded by `cfg!(unix)`, so this is impossible to reach.

weihanglo replied:
Yeah, I just feel like if we don't have this, rustc will yell at us that `arg_max` is not defined. Any better way to rearrange this patch?

jieyouxu replied:
I don't think that is the case? We have a match with two branches, one of them will definitely assign to `arg_max`, the other diverges and so the following will never be reached. At least, I can bootstrap locally without needing this `cfg(not(unix))`-gated `arg_max = 0` assignment.

jieyouxu commented on compiler/rustc_codegen_ssa/src/back/command.rs:
```diff
@@ -137,44 +137,73 @@ impl Command {
+                    .saturating_add(nul)
+                    .saturating_add(ptr_size)
+            });
+            #[cfg(unix)]
```
Nit: redundant, because we're inside the `cfg!(unix)`-guarded branch already :D

weihanglo replied:
https://github.com/rust-lang/rust/pull/138439#issuecomment-2720024265

It turns out not 🤯

`cfg!(…)` only evaluate the cfg value I believe. It doesn't guard the block.

jieyouxu replied:
No you're right, I'm a goober and forgor it's run-time guarded, but `libc` must only be used under `linux` lol.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13827573947/job/38685270827) [(plain)](https://github.com/rust-lang/rust/commit/e8eac7b07ddd9d2213d3b545a9e77a2c818f7fda/checks/38685270827/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_ssa v0.0.0 (/checkout/compiler/rustc_codegen_ssa)
error[E0425]: cannot find value `arg_max` in this scope
   --> compiler/rustc_codegen_ssa/src/back/command.rs:203:51
    |
203 |             args_size.saturating_add(envs_size) > arg_max
    |                                                   ^^^^^^^ not found in this scope

    Checking rustc_privacy v0.0.0 (/checkout/compiler/rustc_privacy)
    Checking rustc_sanitizers v0.0.0 (/checkout/compiler/rustc_sanitizers)
    Checking rustc_mir_build v0.0.0 (/checkout/compiler/rustc_mir_build)
```

</details>

jieyouxu commented:
I would consider restructuring this bit like

```rs
#[cfg(windows)]
{
    // but make sure this returns
}

#[cfg(unix)]
{
    // but make sure this returns
}

false
```

arielb1 commented:
Helping @weihanglo work on this

bors commented:
:umbrella: The latest upstream changes (presumably #138480) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

weihanglo committed:
fix(linker): use arg list estimate on only Windows

Though I doubt anyone running rustc outside Unix/Windows

weihanglo committed:
feat(linker): check ARG_MAX on Unix platforms

On Unix the limits can be gargantuan anyway so we're pretty
unlikely to hit them, but might still exceed it.
We consult ARG_MAX here to get an estimate.

jieyouxu reviewed (approved):
Thanks!

jieyouxu commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 08166b5b239beb794a48bec52adc6e5261b166d1 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 08166b5b239beb794a48bec52adc6e5261b166d1 -->
<!-- homu: {"type":"Approved","sha":"08166b5b239beb794a48bec52adc6e5261b166d1","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
